appservice/src/TunnelProxy.ts

```diff
@@ -0,0 +1,184 @@
+        this._socket.pause();
+
+        this._socket.on('data', (data: Buffer) => {
+            this._outputChannel.appendLine('[Proxy Server] socket data');
+            if (this._wsConnection) {
+                this._wsConnection.send(data);
+            }
+        });
+
+        this._socket.on('close', () => {
+            this._outputChannel.appendLine(`[Proxy Server] client disconnected ${this._socket.remoteAddress}:${this._socket.remotePort}`);
+            this.dispose();
+            this.emit('close');
+        });
+
+        this._socket.on('error', (err: Error) => {
+            this._outputChannel.appendLine(`[Proxy Server] socket error: ${err}`);
+            this.dispose();
+            this.emit('error', err);
+        });
+
+        this._wsClient.on('connect', (connection: websocket.connection) => {
+            this._outputChannel.appendLine('[WebSocket] client connected');
+            this._wsConnection = connection;
+
+            // Resume socket after connection
+            this._socket.resume();
+
+            connection.on('close', () => {
+                this._outputChannel.appendLine('[WebSocket] client closed');
+                this.dispose();
+                this.emit('close');
+            });
+
+            connection.on('error', (err: Error) => {
+                this._outputChannel.appendLine(`[WebSocket error] ${err}`);
+                this.dispose();
+                this.emit('error', err);
+            });
+
+            connection.on('message', (data: websocket.IMessage) => {
+                this._outputChannel.appendLine('[WebSocket] data');
+                this._socket.write(data.binaryData);
+            });
+
+        });
+
+        this._wsClient.on('connectFailed', (err: Error) => {
+            this._outputChannel.appendLine(`[WebSocket] connectFailed: ${err}`);
+            this.dispose();
+            this.emit('error', err);
+        });
+
+        this._wsClient.connect(
+            `wss://${this._client.kuduHostName}/AppServiceTunnel/Tunnel.ashx`,
+            undefined,
+            undefined,
+            { 'Cache-Control': 'no-cache', Pragma: 'no-cache' },
+            { auth: `${this._publishCredential.publishingUserName}:${this._publishCredential.publishingPassword}` }
+        );
+    }
+
+    public dispose(): void {
+        this._outputChannel.appendLine('[Proxy Server] socket dispose');
+
+        if (this._wsConnection) {
+            this._wsConnection.close();
+            this._wsConnection = undefined;
+        }
+
+        if (this._wsClient) {
+            this._wsClient.abort();
+            this._wsClient = undefined;
+        }
+
+        if (this._socket) {
+            this._socket.destroy();
+            this._socket = undefined;
+        }
+    }
+}
+
+/**
+ * A local TCP server that forwards all connections to the Kudu tunnel websocket endpoint.
+ */
+export class TunnelProxy {
```

Comment:
I assume this is only used for remote debug (correct me if I'm wrong), in which case "TunnelProxy" feels too generic. What about something like "RemoteDebugProxy"?

Reply:
Okay in that case I guess I don't have a strong preference